internal/postgres/process.go

```diff
@@ -0,0 +1,163 @@
+		conv.statsAddBadRow(spTable, conv.dataMode())
+		r := &row{table: spTable, cols: cols, vals: vals}
+		bytes := byteSize(r)
+		// Cap storage used by badRows. Keep at least one bad row.
+		if len(conv.sampleBadRows.rows) == 0 || bytes+conv.sampleBadRows.bytes < conv.sampleBadRows.bytesLimit {
+			conv.sampleBadRows.rows = append(conv.sampleBadRows.rows, r)
+			conv.sampleBadRows.bytes += bytes
+		}
+	} else {
+		if conv.dataSink == nil {
+			msg := "Internal error: ProcessRow called but dataSink not configured"
+			internal.VerbosePrintf("%s\n", msg)
+			conv.unexpected(msg)
+			conv.statsAddBadRow(spTable, conv.dataMode())
+		} else {
+			conv.dataSink(spTable, c, v)
+			conv.statsAddGoodRow(spTable, conv.dataMode())
+		}
+	}
+}
+
+func byteSize(r *row) int64 {
+	n := int64(len(r.table))
+	for _, c := range r.cols {
+		n += int64(len(c))
+	}
+	for _, v := range r.vals {
+		n += int64(len(v))
+	}
+	return n
+}
+
+// readAndParseChunk parses a chunk of pg_dump data, returning the bytes read,
+// the parsed AST (nil if nothing read), and whether we've hit end-of-file.
+func readAndParseChunk(conv *Conv, r *internal.Reader) ([]byte, []nodes.Node) {
+	var l [][]byte
+	for {
+		b := r.ReadLine()
+		l = append(l, b)
+		// If we see a semicolon or eof, we're likely to have a command, so try to parse it.
+		// Note: we could just parse every iteration, but that would mean more attempts at parsing.
+		if strings.Contains(string(b), ";") || r.EOF {
+			n := 0
+			for i := range l {
+				n += len(l[i])
+			}
+			s := make([]byte, n)
+			n = 0
+			for i := range l {
+				n += copy(s[n:], l[i])
+			}
+			tree, err := pg_query.Parse(string(s))
+			if err == nil {
+				return s, tree.Statements
+			}
+			// Likely causes of failing to parse:
+			// a) complex statements with embedded semicolons e.g. 'CREATE FUNCTION'
+			// b) a semicolon embedded in a multi-line comment, or
+			// c) a semicolon embedded a string constant or column/table name.
+			// We deal with this case by reading another line and trying again.
+			conv.stats.reparsed++
+		}
+		if r.EOF {
+			if len(l) != 0 {
+				fmt.Printf("Error parsing last %d line(s) of input\n", len(l))
+			}
+			return nil, nil
+		}
+	}
+}
+
+func processCopyBlock(conv *Conv, c *copyOrInsert, r *internal.Reader) {
+	internal.VerbosePrintf("Parsing COPY-FROM stdin block starting at line=%d/fpos=%d\n", r.LineNumber, r.Offset)
+	for {
+		b := r.ReadLine()
+		if string(b) == "\\.\n" {
+			internal.VerbosePrintf("Parsed COPY-FROM stdin block ending at line=%d/fpos=%d\n", r.LineNumber, r.Offset)
+			return
+		}
+		if r.EOF {
+			conv.unexpected("Reached eof while parsing copy-block")
+			return
+		}
+		conv.statsAddRow(c.spTable, conv.schemaMode())
+		// We have to read the copy-block data so that we can process the remaining
+		// pg_dump content. However, if we don't want the data, stop here.
+		// In particular, avoid the strings.Split and ProcessRow calls below, which
+		// weill be expensive for huge datasets.
```

Comment:
typo s/weill/will/
I am assuming that there is not schema after copy-block hence you can stop processing if it is only schema mode.

Reply:
Done.

Yes, that's right. The copy-block just contains data, so we skip it in schema mode.

But we do have to start processing again after the end of the copy-block to process any statements that follow the copy-block (which can contain more scheme info).